hera_cal/tempcal.py

```diff
@@ -0,0 +1,203 @@
+class TempCal(VisClean):
+    
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize TempCal, see vis_clean.VisClean for kwargs.
+        """ 
+        super(TempCal, self).__init__(*args, **kwargs)
+
+        # init empty dictionaries
+        self.gains = {}
+        self.gflags = {}
+
+    def gains_from_autos(self, data, times, flags=None, smooth_frate=1.0, nl=1e-10, Nmirror=0,
+                         keys=None, edgeflag=0, verbose=True):
+        """
+        Model temperature fluctuations in auto-correlations
+        by dividing auto-correlation by its time-smoothed counterpart.
+
+        This is only valid for timescales significantly shorter than
+        the beam crossing timescale. Resultant gains have a "divide" gain_convention.
+
+        Rcommended to factorize the flags beforehand using utils.factorize_flags.
+
+        Args:
+            data : DataContainer
+                Holds auto-correlation waterfalls of shape (Ntimes, Nfreqs)
+            times : ndarray
+                Holds time array in Julian Date of shape (Ntimes,)
+            flags : DataContainer
+                Holds data flags of shape (Ntimes, Nflags)
+            smooth_frate : float
+                Fringe-Rate cutoff for smoothing in mHz
+            nl : float
+                Noise level in GP interpolation.
+            Nmirror : int
+                Number of time bins to mirror about edges before interpolation
+            keys : list
+                List of ant-pair-pol tuples to operate on
+            edgeflag : int or len-2 tuple
+                Number of channels to flag on low and high end of frequency band.
+                Low and high can be specified independently via a len-2 tuple.
+            verbose : bool
+                If True, report feedback to stdout.
+        """
+        # get flags
+        if flags is None:
+            flags = DataContainer(dict([(k, np.zeros_like(data[k], np.bool)) for k in data]))
+
+        # edgeflag
+        if edgeflag is not None:
+            flags = copy.deepcopy(flags)
+            if isinstance(edgeflag, (int, np.int, float, np.float)):
+                edgeflag = (edgeflag, edgeflag)
+            assert len(edgeflag) == 2
+            for k in flags:
+                if edgeflag[0] > 0:
+                    flags[k][:, :edgeflag[0]] = True
+                if edgeflag[1] > 0:
+                    flags[k][:, -edgeflag[1]:] = True
+
+        # get length scale in seconds
+        length_scale = 1.0 / (smooth_frate * 1e-3) / (24.0 * 3600.0)
+
+        # get keys
+        if keys is None:
+            keys = list(data.keys())
+
+        # smooth data
+        if not hasattr(self, 'smooth'):
+            self.smooth = DataContainer({})
+        if not hasattr(self, 'ratio'):
+            self.ratio = DataContainer({})
+        for k in keys:
+            # smooth data
+            utils.echo("working on {}".format(k), verbose=verbose)
+            gkey = utils.split_bl(k)[0]
```

Comment:
This makes an assumption that `utils.split_bl(k)[0] == utils.split_bl(k)[1]`, i.e. that it's an autocorrelation. You might want to enforce that assumption with an assert.

Reply:
yup sure thing